BrainPortal/app/controllers/data_providers_controller.rb

```diff
@@ -110,6 +115,39 @@ def create #:nodoc:
     end
   end
 
+  # dp creation by normal user, currently only UserkeyFlatDirSshDataProvider
+  # allow user create his own data providers.  At the moment only ssh with UserkeyFlatDirSshDataProvider and few field only
+  def create_by_normal_user
+
+    normal_params = params.require_as_params(:data_provider)
+                        .permit(:name, :description, :group_id,
+                                :remote_user, :remote_host,
+                                :remote_port, :remote_dir
+                                )
+    group_id = normal_params[:date_provider][:group_id]
+    Group.where(id: current_user.assignable_group_ids).find(group_id) # ensure assignable, not sure need check visibility etc more
```

Comment:
Why are you querying the DB twice? Why not instead:

```ruby
current_user.assignable_groups.find(group_id)
```

It's not for nothing that we added all these utility methods to the `User` class for fetching groups!